sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,635 @@
+
+  AddHeaders(responseHeaders, rawResponse);
+
+  int64_t contentLength
+      = GetContentLength(handleManager, requestMethod, rawResponse->GetStatusCode());
+
+  rawResponse->SetBodyStream(std::move(GetBodyStream(handleManager, contentLength)));
+  return rawResponse;
+}
+
+std::unique_ptr<RawResponse> WinHttpTransport::Send(Context const& context, Request& request)
+{
+  // TODO: Need to test if context has been canceled and cleanup, can't call ThrowIfCanceled.
+  (void)(context);
+
+  auto handleManager = std::make_unique<Details::HandleManager>(context, request);
+
+  GetSessionHandle(handleManager);
+  GetConnectionHandle(handleManager);
+  GetRequestHandle(handleManager);
+
+  SendRequest(handleManager);
+
+  ReceiveResponse(handleManager);
+
+  return GetRawResponse(handleManager, request.GetMethod());
+}
+
+// Read the response from the sent request.
+int64_t Details::WinHttpStream::Read(Context const& context, uint8_t* buffer, int64_t count)
+{
+  context.ThrowIfCanceled();
+
+  if (count <= 0 || this->m_isEOF)
+  {
+    return 0;
+  }
+
+  DWORD numberOfBytesRead = 0;
+
+  // Check for available data.
+  DWORD numberOfBytesAvailable = 0;
+  if (!WinHttpQueryDataAvailable(this->m_handleManager->m_requestHandle, &numberOfBytesAvailable))
+  {
+    // Errors include:
+    // ERROR_WINHTTP_CONNECTION_ERROR
+    // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_TIMEOUT
+    // ERROR_NOT_ENOUGH_MEMORY
+
+    DWORD error = GetLastError();
+    throw Azure::Core::Http::TransportException(
+        "Error while querying how much data is available to read. Error Code: "
+        + std::to_string(error) + ".");
+  }
+
+  context.ThrowIfCanceled();
+
+  DWORD numberOfBytesToRead = numberOfBytesAvailable;
+  if (numberOfBytesAvailable > count)
+  {
+    numberOfBytesToRead = static_cast<DWORD>(count);
+  }
+
+  if (!WinHttpReadData(
+          this->m_handleManager->m_requestHandle,
+          (LPVOID)(buffer),
+          numberOfBytesToRead,
+          &numberOfBytesRead))
+  {
+    // Errors include:
+    // ERROR_WINHTTP_CONNECTION_ERROR
+    // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_RESPONSE_DRAIN_OVERFLOW
+    // ERROR_WINHTTP_TIMEOUT
+    // ERROR_NOT_ENOUGH_MEMORY
+
+    DWORD error = GetLastError();
+    throw Azure::Core::Http::TransportException(
+        "Error while querying how much data is available to read. Error Code: "
+        + std::to_string(error) + ".");
+  }
+
+  this->m_streamTotalRead += numberOfBytesRead;
 
-  throw;
+  if (numberOfBytesRead == 0
+      || (this->m_contentLength != -1 && this->m_streamTotalRead == this->m_contentLength))
+  {
+    this->m_isEOF = true;
```

Comment:
Instead of setting this state, I would save the `numberOfBytesRead` as something like `lastRead`.
It seems to me a more valuable information to have.
Then, on the next call, of the `lastRead ==0` you consider that as EOF().

This is part of the same approach of adding information to the state and not extra state variables describing the current state.

Reply:
How is saving `lastRead` in the state heling us reduce the state being tracked? It seems we are just replacing one with the other.

Reply:
It adds a meaningful and independent information to the state.
lastRead can tell:
- When you are reading for the first time (set to -1 for instance)
- If last read was 0 (eof)

The problem with the m_isEOF is that it represent a second layer of state. It needs to be set after some specific meaningful state has detected, so you introduce the detecting that state code some places and then some other places until it becomes a mess.

Adding information to the state is not bad when it is something that describes directly to the component.

Here's a silly example
For a Dog class, it would be unlikely to add an internal state `m_hasMoved`  that you would update whenever another property like `m_position` changes. Because position can change from many places. The dog can walk, can be pushed, can run, can jump, etc.  In all those actions, the position change and then you will need to update m_hasMoved.
In the contrary, when we have a method HasMoved() that check if the position has changed and return true, it doesn't matter if the position keeps changing from many places, the method will use the state to tell about it only when it is required to know about it

Reply:
Based on offline conversation, leaving it as is, since this is an implementation detail choice and functionally correct.